### PR TITLE
Simplify `hostdevice_vector`

### PR DESCRIPTION
**Pull request by upsj**
## Description
The implementation of `hostdevice_vector` has a lot of code duplication and a defect (probably) because its span cast operators use `max_size` and not `size`. I'd like to give `hostdevice_vector` more `thrust::host_vector`-like functionality and semantics, so this is a first step towards that goal.

Needs to be benchmarked before merging, to see if the default-initialization slows anything down significantly

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — CHANGES_REQUESTED** [review]
@upsj This is a great refactor. I've been meaning to do this for so long and left myself [a little note](https://github.com/rapidsai/cudf/blob/2552c4c97085771a5c6c5ade317aaee28ab7bae3/cpp/src/io/utilities/hostdevice_vector.hpp#L165) to fix it!! I have a few comments.

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 1]
```diff
@@ -37,71 +40,52 @@ class hostdevice_vector {
...
-  bool insert(const T& data)
+  void insert(const T& data)
   {
-    if (num_elements < max_elements) {
-      h_data[num_elements] = data;
-      num_elements++;
-      return true;
-    }
-    return false;
+    CUDF_EXPECTS(size() < capacity(), "hostdevice_vector overflow");
```
Comment (bdice) [inline cpp/src/io/utilities/hostdevice_vector.hpp:61]: This error can probably be more helpful to the user.
```suggestion
    CUDF_EXPECTS(size() < capacity(), "Cannot insert data into hostdevice_vector because max_size has been exceeded.");
```

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 2]
```diff
@@ -37,71 +40,52 @@ class hostdevice_vector {
...
-  }
-
-  ~hostdevice_vector()
+    : d_data(0, stream)
   {
-    if (max_elements != 0) {
-      [[maybe_unused]] auto const free_result = cudaFreeHost(h_data);
-      assert(free_result == cudaSuccess);
-    }
+    CUDF_EXPECTS(initial_size <= max_size, "initial_size can't be larger than max_size");
```
Comment (davidwendt) [inline cpp/src/io/utilities/hostdevice_vector.hpp:53]:
```suggestion
    CUDF_EXPECTS(initial_size <= max_size, "initial_size cannot be larger than max_size");
```

**Comment by upsj** [comment]
Maybe I should note that one of the impacts of this change is that the two-parameter constructor now causes all elements to be default-initialized, while they were uninitialized before. This might have a small performance impact on the host-side?

**Review by bdice — APPROVED** [review]
Nice work. I think this is a helpful refactoring. It'd be good to check performance if you think the default initialization of the host vector will matter, but I don't anticipate any issues there.

**Review by davidwendt — APPROVED** [review]

File: cpp/src/io/parquet/reader_impl.cu [hunk 3]
```diff
@@ -1637,13 +1637,6 @@ table_with_metadata reader::impl::read(std::vector<std::vector<size_type>> const
         // this column contains repetition levels and will require a preprocess
         if (schema.max_repetition_level > 0) { has_lists = true; }
 
-        // Spec requires each row group to contain exactly one chunk for every
-        // column. If there are too many or too few, continue with best effort
-        if (chunks.size() >= chunks.max_size()) {
-          std::cerr << "Detected too many column chunks" << std::endl;
-          continue;
-        }
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl.cu:1645]: saw this one too, was gonna open a PR :D

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 4]
```diff
@@ -37,71 +40,53 @@ class hostdevice_vector {
...
-      assert(free_result == cudaSuccess);
-    }
+    CUDF_EXPECTS(initial_size <= max_size, "initial_size cannot be larger than max_size");
+    h_data.reserve(max_size);
+    h_data.resize(initial_size);
+    d_data.resize(max_size, stream);
   }
 
-  bool insert(const T& data)
+  void insert(const T& data)
```
Comment (vuule) [inline cpp/src/io/utilities/hostdevice_vector.hpp:59]: Going along with host_vector API:
```suggestion
  void push_back(const T& data)
```

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 5]
```diff
@@ -37,71 +40,53 @@ class hostdevice_vector {
...
-    : max_elements(max_size), num_elements(initial_size)
-  {
-    if (max_elements != 0) {
-      CUDF_CUDA_TRY(cudaMallocHost(reinterpret_cast<void**>(&h_data), sizeof(T) * max_elements));
-      d_data.resize(sizeof(T) * max_elements, stream);
-    }
-  }
-
-  ~hostdevice_vector()
+    : d_data(0, stream)
```
Comment (vuule) [inline cpp/src/io/utilities/hostdevice_vector.hpp:51]: why is this needed?

**Review by vuule — COMMENTED** [review]
Great changes!
Just a `std::pair<question, suggestion>`

File: cpp/src/io/utilities/hostdevice_vector.hpp (hunk 4 above)
Comment (bdice) [reply cpp/src/io/utilities/hostdevice_vector.hpp:59]: Ah, good eye. I was going to comment on this and forgot.

File: cpp/src/io/utilities/hostdevice_vector.hpp (hunk 5 above)
Comment (upsj) [reply cpp/src/io/utilities/hostdevice_vector.hpp:51]: device_uvector doesn't have a default constructor, so we need to provide a stream

**Comment by upsj** [comment]
I ran the FST, Parquet reader and writer benchmarks and found no significant performance differences.

**Comment by upsj** [comment]
rerun tests

**Comment by Ethyling** [comment]
rerun tests

**Comment by upsj** [comment]
@gpucibot merge
